kube/src/client/mod.rs

```diff
@@ -130,35 +131,60 @@ impl Client {
-            // - No host found in URL
-            // - Unsupported scheme (not ws/wss)
-            // shouldn't happen in our case
-            Err(ws2::Error::Url(msg)) => Err(Error::RequestValidation(msg.into())),
-
-            // Protocol errors:
-            // - Only GET is supported
-            // - Only HTTP version >= 1.1 is supported
-            // shouldn't happen in our case
-            Err(ws2::Error::Protocol(msg)) => Err(Error::RequestValidation(msg.into())),
-
-            // Fatal error with undelying connection.
-            Err(ws2::Error::Io(err)) => panic!("WebSocket connection error: {}", err),
-
-            // Unknown error. `tungstenite::Error` includes those that only happens after connecting.
-            Err(err) => panic!("Unknown WebSocket error: {}", err),
+            Err(err) => match err {
+                // tungstenite only gives us the status code.
+                ws2::Error::Http(code) => Err(Error::Api(ErrorResponse {
+                    status: code.to_string(),
+                    code: code.as_u16(),
+                    message: "".to_owned(),
+                    reason: "".to_owned(),
+                })),
+
+                ws2::Error::HttpFormat(err) => Err(Error::HttpError(err)),
+
+                // `tungstenite::Error::Tls` is only available when using `ws-native-tls` (`async-tungstenite/tokio-native-tls`)
+                // because it comes from `tungstenite/tls` feature.
+                #[cfg(feature = "ws-native-tls")]
+                ws2::Error::Tls(err) => Err(Error::SslError(format!("{}", err))),
+
+                // URL errors:
+                // - No host found in URL
+                // - Unsupported scheme (not ws/wss)
+                // shouldn't happen in our case
+                ws2::Error::Url(msg) => Err(Error::RequestValidation(msg.into())),
+
+                // Protocol errors:
+                // - Only GET is supported
+                // - Only HTTP version >= 1.1 is supported
+                // shouldn't happen in our case
+                ws2::Error::Protocol(msg) => Err(Error::RequestValidation(msg.into())),
+
+                ws2::Error::Io(err) => match err.kind() {
+                    std::io::ErrorKind::AddrNotAvailable
+                    | std::io::ErrorKind::ConnectionRefused
+                    | std::io::ErrorKind::ConnectionReset
+                    | std::io::ErrorKind::ConnectionAborted
+                    | std::io::ErrorKind::Interrupted
+                    | std::io::ErrorKind::InvalidInput
+                    | std::io::ErrorKind::InvalidData
+                    | std::io::ErrorKind::NotConnected
+                    | std::io::ErrorKind::NotFound
+                    | std::io::ErrorKind::TimedOut
+                    | std::io::ErrorKind::UnexpectedEof
+                    | std::io::ErrorKind::WouldBlock => Err(Error::Connection(err)),
+
+                    _ => panic!("WebSocket connection error: {}", err),
```

Comment:
For other stuff, I think we'll need to match with `std::io::ErrorKind::Other` and stringify the error, then test if the string contains errors we'd like to avoid panic...

Reply:
Could we not just put all the io errors into `Error::Connection`? I feel `panic!` is the wrong default for error cases we didn't cover.

Reply:
Sure, that's easy :)